tests/zkevm/test_worst_memory.py

```diff
@@ -0,0 +1,262 @@
+    )
+
+
+@pytest.mark.valid_from("Cancun")
+@pytest.mark.parametrize(
+    "max_code_size_ratio",
+    [
+        pytest.param(0, id="0 bytes"),
+        pytest.param(0.25, id="0.25x max code size"),
+        pytest.param(0.50, id="0.50x max code size"),
+        pytest.param(0.75, id="0.75x max code size"),
+        pytest.param(1.00, id="max code size"),
+    ],
+)
+@pytest.mark.parametrize("fixed_dst", [True, False])
+def test_worst_codecopy(
+    state_test: StateTestFiller,
+    pre: Alloc,
+    fork: Fork,
+    max_code_size_ratio: float,
+    fixed_dst: bool,
+):
+    """Test running a block filled with CODECOPY executions."""
+    env = Environment()
+    max_code_size = fork.max_code_size()
+
+    size = int(max_code_size * max_code_size_ratio)
+
+    code_prefix = Op.PUSH32(size)
+    jumpdest = Op.JUMPDEST
+    jump_back = Op.JUMP(len(code_prefix))
+    dst = 0 if fixed_dst else Op.MOD(Op.GAS, 7)
+    iter_block = Op.CODECOPY(dst, Op.PUSH0, Op.DUP1)
+    max_iters_loop = (max_code_size - len(code_prefix) - len(jumpdest) - len(jump_back)) // len(
+        iter_block
+    )
+    code = code_prefix + jumpdest + sum([iter_block] * max_iters_loop) + jump_back
+    if len(code) > max_code_size:
+        raise ValueError(f"Code size {len(code)} exceeds maximum code size {max_code_size}")
+
+    tx = Transaction(
+        to=pre.deploy_contract(code=code),
+        gas_limit=env.gas_limit,
+        data=b"\xff" * size,
+        sender=pre.fund_eoa(),
+    )
+
+    state_test(
+        genesis_environment=env,
+        pre=pre,
+        post={},
+        tx=tx,
+    )
+
+
+@pytest.mark.valid_from("Cancun")
+@pytest.mark.parametrize(
+    "size",
+    [
+        pytest.param(0, id="0 bytes"),
+        pytest.param(10, id="10 bytes"),
+        pytest.param(100, id="100 bytes"),
+        pytest.param(1 * 1024, id="1KiB"),
+        pytest.param(10 * 1024, id="10KiB"),
+        pytest.param(100 * 1024, id="100KiB"),
+        pytest.param(1024 * 1024, id="1MiB"),
+    ],
+)
+@pytest.mark.parametrize("fixed_dst", [True, False])
+def test_worst_returndatacopy(
+    state_test: StateTestFiller,
+    pre: Alloc,
+    fork: Fork,
+    size: int,
+    fixed_dst: bool,
+):
+    """Test running a block filled with RETURNDATACOPY executions."""
+    env = Environment()
+    max_code_size = fork.max_code_size()
+
+    # Create the contract that will RETURN the data that will be used for RETURNDATACOPY.
+    # Random-ish data is injected at different points in memory to avoid making the content
+    # predictable. If `size` is 0, this helper contract won't be used.
+    code = (
+        Op.MSTORE8(0, Op.GAS)
+        + Op.MSTORE8(size // 2, Op.GAS)
+        + Op.MSTORE8(size - 1, Op.GAS)
+        + Op.RETURN(0, size)
+    )
+    helper_contract = pre.deploy_contract(code=code)
+
+    # We create the contract that will be doing the RETURNDATACOPY multiple times.
+    code_prefix = Op.STATICCALL(address=helper_contract) if size > 0 else Bytecode()
```

Comment:
For `size == 0` let's avoid the extra overhead, plus the above `MSTORE8`-3-byte-strategy doesn't make sense either.